applications/rate_limiting_nullifier.md

```diff
@@ -0,0 +1,255 @@
+| -----: | ----------- | ------------- |
+| **0a.** | License | Apache 2.0 / GPLv3 / MIT / Unlicense |
+| **0b.** | Documentation | We will provide both **inline documentation** of the code and a basic **tutorial** that explains how a user can (for example) spin up one of our Substrate nodes and send test transactions, which will show how the new functionality works. |
+| **0c.** | Testing and Testing Guide | Core functions will be fully covered by comprehensive unit tests to ensure functionality and robustness. In the guide, we will describe how to run these tests. |
+| **0d.** | Docker | We will provide a Dockerfile(s) that can be used to test all the functionality delivered with this milestone. |
+| 1. | Multiwriter Event Feed Pallet | We will implement a pallet which allows for emitting events conformant to the spec, and based on RLN proof verification  |
+| 2. | Multiwriter Event Feed Pallet | The pallet will allow for generating events dependent on one another |
+| 3. | Multiwriter Event Feed Pallet | The pallet will allow for locking/reserving some amount for participation in semaphore or RLN group |
+| 4. | Multiwriter Event Feed Pallet | The pallet will contain an example for accepting proof verification as an alternative to signed transaction verification |
+### Milestone 4 Adapt to alternative proof system
+
+- **Estimated Duration:** 2 Months
+- **FTE:** 2
+- **Costs:** $3000 USD
+
+| Number | Deliverable | Specification |
+| -----: | ----------- | ------------- |
+| **0a.** | License | Apache 2.0 / GPLv3 / MIT / Unlicense |
+| **0b.** | Documentation | We will provide both **inline documentation** of the code and a basic **tutorial** that explains how a user can (for example) spin up one of our Substrate nodes and send test transactions, which will show how the new functionality works. |
+| **0c.** | Testing and Testing Guide | Core functions will be fully covered by comprehensive unit tests to ensure functionality and robustness. In the guide, we will describe how to run these tests. |
+| **0d.** | Docker | We will provide a Dockerfile(s) that can be used to test all the functionality delivered with this milestone. |
+| 0e. | Article | We will publish an article for Substrate developers that explains the significance of the pallets, the foundations of Semaphore and Rate-Limiting Nullifier, their applicability to different types of applications, and how to incorporate them in Substrate-based chains |
+| 1. | Plonky2 Circuit | We will write the Semaphore circuit using a plonky2 library |
+| 2. | Plonky2 Verifier | We will implement a verifier pallet primitive for Plonky2 to verify proofs from the new Plonky2 circuit |
+| 3. | Plonky2 Circuit | We will implement the circuit inputs and constraints for Rate-Limiting Nullifier |
+
+
+## Future Plans
+- Documentation in the form of gitbook showing terminology, context, introduction, tutorial, and basic example usage of the delivered pallets for relevant use cases.
```

Comment:
We're just excited in the ZK ecosystem, and want to build on it in Substrate. This would mainly be for sharing and learning. We like to experiment with similar ideas, so there's a chance we may build on it in the future when working on something similar.